openfisca_core/entities.py

```diff
@@ -441,7 +474,10 @@ def nb_persons(self, role = None):
             If ``role`` is provided, only the entity member with the given role are taken into account.
         """
         if role:
-            role_condition = self.members.has_role(role)
```

Comment:
Not sure I understand the reason of this change. Why can't we keep using the `has_role` method?

Reply:
@fpagnoux We are refactoring towards code that obeys the Law of Demeter, to reduce the entanglement of dependencies which, to me, is a main component of our technical debt.

The state of play before this PR is that `nb_persons` of `GroupEntity` is implemented via `has_role` of `PersonEntity` (for the branch where a role is supplied). This creates a dependency `GroupEntity -> PersonEntity`. But `has_role` of `PersonEntity` is implemented via a dependency on `Simulation`, to get the `GroupEntity` corresponding to the role, and finally to access `entity.members_role` which should be an internal implementation detail of GroupEntity. We have created a dependency `PersonEntity -> Simulation -> GroupEntity`.

Composing these dependencies we get `GroupEntity -> PersonEntity -> Simulation -> GroupEntity`, a circular dependency. This is one of the reasons, based on my experience, why things are hard to unit test.

After this PR, `nb_persons` can be called on a `GroupEntity` and do its job without relying on these collaborators, using exclusively the information it has and which it owns, `members_role`. This is easy to unit test, more maintainable.

I think it should also be possible to improve `has_role` (because of the dependency on `Simulation`, which seems very awkward to me) but that's a later topic.

Reply:
👌 
The duplication of a non-trivial code bit seems a little unfortunate to me though. 
Could we wrap-up the `np.logical_or.reduce` in a group-entity method, so that both `has_role`  and `nb_persons` can use it?  

Reply:
Short answer: no. Part of the problem is that `has_role` is a `PersonEntity` method, while `nb_persons` is a `GroupEntity` method. The data that is being reduced belongs to `GroupEntity`.

I can't see a way to solve this without first addressing the problems in `has_role` - specifically the fact that it requires access to all group entities via the Simulation instance.

This is partially a consequence of the signature for formulas - you get a `person` but not the other entities so that `person` is the only thing you can call `has_role` on. And implicitly, `has_role` takes **two** arguments, not one; you don't ask "what role a person has" generically, you're always asking "what role does a person fill in a particular group". The problem is that the Role values are too lightweight (they're [flyweights in the GoF patterns sense](https://en.wikipedia.org/wiki/Flyweight_pattern)) to carry this information.

So the underlying question is, what better approach can we find for filtering by role, and how does this affect the signature of formulas… which is way out of scope for this PR.

Reply:
> As much as I appreciate technical debt reduction and better adhesion to OOP good practices in Core, I'd be worried is this was the main driver to big changes on something as frontal to users as the formulas signature.
> Ease of use for users (including tech-savy economists that are not dev) and expressivity of the formulas (to be assessed by user tests and user feedback) is IMHO the most important aspect of formula syntax.

Reply:
> As much as I appreciate technical debt reduction and better adhesion to OOP good practices in Core, I'd be worried is this was the main driver to big changes on something as frontal to users as the formulas signature.
> Ease of use for users (including tech-savy economists that are not dev) and expressivity of the formulas (to be assessed by user tests and user feedback) is IMHO the most important aspect of formula syntax.